File: $_algotsu-kaizen/problemo-set/binary-tree-postorder-traversal.c
/**
 * 145. Binary Tree Postorder Traversal
 * https://leetcode.com/problems/binary-tree-postorder-traversal/description/
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int countNodes(struct TreeNode* root) {
    if (root == NULL) return 0b0;
    return 0b1 + countNodes(root->left) + countNodes(root->right);
}

int* recusor(struct TreeNode* root, int* accumulator, int* idx) {
    if (root == NULL) return accumulator;
    if (root->left != NULL) recusor(root->left, accumulator, idx);
    if (root->right != NULL) recusor(root->right, accumulator, idx);
    *(accumulator + (*idx)++) = root->val;
    return accumulator;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* postorderTraversal(struct TreeNode* root, int* returnSize) {
    int nodeCount = countNodes(root);
    int index = 0;
    *returnSize = nodeCount;
    return recusor(root, (int *)malloc(sizeof(int) * nodeCount), &index);
}

File: $_algotsu-kaizen/problemo-set/binary-tree-postorder-traversal.js
/**
 * 145. Binary Tree Postorder Traversal
 * https://leetcode.com/problems/binary-tree-postorder-traversal/description/
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function(root, accumulator = []) {
    root.left && postorderTraversal(root.left, accumulator)
    root.right && postorderTraversal(root.right, accumulator)
    accumulator.push(root.val)
    return accumulator
};

export { postorderTraversal }

// Input: n = "123"
// Output: "121"
console.log("$#",
    postorderTraversal(
        "1234555"
    ),
    "121"
);

// Input: n = "1"
// Output: "0"
console.log("$#",
    postorderTraversal(
        "1"
    ),
    "0"
);


// 1234, 999, 1000
console.log("$#",
    postorderTraversal(
        "1234"
    ),
    "1221"
);
